rig/place_and_route/util.py

```diff
@@ -0,0 +1,105 @@
+"""Utilities for common-case usage of place-and-route facilities.
+"""
+
+from collections import defaultdict, deque
+
+from six import iteritems
+
+from ..machine import Links, Cores
+
+from ..routing_table import Routes, RoutingTableEntry
+
+from .routing_tree import RoutingTree
+
+
+def build_application_map(vertices_applications, placements, allocations,
+                          core_resource=Cores):
+    """Build a mapping from application to a list of cores where the
+    application is used.
+
+    This utility function assumes that each vertex is associated with a
+    specific application.
+
+    Arguments
+    ---------
+    vertices_applications : {vertex: application, ...}
+        Applications are represented by the path of their APLX file.
+    placements : {vertex: (x, y), ...}
+    allocations : {vertex: {resource: slice, ...}, ...}
+        One of these resources should match the `core_resource` argument.
+    core_resource : object
+        The resource identifier which represents cores.
+
+    Returns
+    -------
+    {application: {(x, y) : set([c, ...]), ...}, ...}
+        For each application, for each used chip a set of core numbers onto
+        which the application should be loaded.
+    """
+    application_map = defaultdict(lambda: defaultdict(set))
+
+    for vertex, application in iteritems(vertices_applications):
+        chip_cores = application_map[application][placements[vertex]]
+        core_slice = allocations[vertex].get(core_resource, slice(0, 0))
+        chip_cores.update(range(core_slice.start, core_slice.stop))
+
+    return application_map
+
+
+def build_routing_tables(routes, net_keys):
+    """Convert a set of RoutingTrees into a per-chip set of routing tables.
+
+    This command produces routing tables with entries ommitted when the route
+    does not change direction.
```

Comment:
I presume this is the case for your minimisation algorithm? If so I'd be happy to add the option :).

Re default routing being turned off, it can't be, minus someone adding a globally matching route for drop packets to the end of the routing table...